### PR TITLE
feat: check for outdated packages 

### PR DESCRIPTION
**Pull request by kaf-lamed-beyt**
Hi @ericallam, kindly review this. 

Although, I haven't been able to test, because my PC went off. When it's back up I'll try testing it by creating an example Next.js app in the `examples` dir

/claim #167

### CODE REVIEW

**Comment by kaf-lamed-beyt** [comment]
While trying to test, I'm stuck at the point where I have to do `pnpm run db:migrate`. Kindly take a look at the error I'm getting. 

```shell
pnpm run db:migrate 

> triggerdotdev@0.0.1 db:migrate /home/seven/Documents/codes/oss/trigger.dev
> turbo run db:migrate:deploy generate

• Packages in scope: @examples/job-catalog, @examples/jobs-starter, @examples/nextjs, @examples/package-tester, @trigger.dev/cli, @trigger.dev/core, @trigger.dev/database, @trigger.dev/express, @trigger.dev/github, @trigger.dev/integration-kit, @trigger.dev/nextjs, @trigger.dev/openai, @trigger.dev/plain, @trigger.dev/react, @trigger.dev/resend, @trigger.dev/sdk, @trigger.dev/slack, @trigger.dev/stripe, @trigger.dev/supabase, @trigger.dev/tailwind-config, @trigger.dev/tsconfig, @trigger.dev/typeform, docs, emails, eslint-config-custom, nextjs-12, outdated-packages, webapp
• Running db:migrate:deploy, generate in 28 packages
• Remote caching disabled
@trigger.dev/database:db:migrate:deploy: cache bypass, force executing 909df10673c0b4f0
@trigger.dev/database:generate: cache miss, executing 147eb0d5f9a081ae
@trigger.dev/database:generate: 
@trigger.dev/database:generate: > @trigger.dev/database@0.0.0 generate /home/seven/Documents/codes/oss/trigger.dev/packages/database
@trigger.dev/database:generate: > prisma generate
@trigger.dev/database:generate: 
@trigger.dev/database:db:migrate:deploy: 
@trigger.dev/database:db:migrate:deploy: > @trigger.dev/database@0.0.0 db:migrate:deploy /home/seven/Documents/codes/oss/trigger.dev/packages/database
@trigger.dev/database:db:migrate:deploy: > prisma migrate deploy
@trigger.dev/database:db:migrate:deploy: 
@trigger.dev/database:generate: Environment variables loaded from .env
@trigger.dev/database:generate: Prisma schema loaded from prisma/schema.prisma
@trigger.dev/database:db:migrate:deploy: Environment variables loaded from .env
@trigger.dev/database:db:migrate:deploy: Prisma schema loaded from prisma/schema.prisma
@trigger.dev/database:db:migrate:deploy: Datasource "db": PostgreSQL database "postgres", schema "public" at "localhost:5432"
@trigger.dev/database:db:migrate:deploy: 
@trigger.dev/database:db:migrate:deploy: Error: P1001: Can't reach database server at `localhost`:`5432`
@trigger.dev/database:db:migrate:deploy: 
@trigger.dev/database:db:migrate:deploy: Please make sure your database server is running at `localhost`:`5432`.
@trigger.dev/database:db:migrate:deploy:  ELIFECYCLE  Command failed with exit code 1.
@trigger.dev/database:db:migrate:deploy: ERROR: command finished with error: command (/home/seven/Documents/codes/oss/trigger.dev/packages/database) pnpm run db:migrate:deploy exited (1)
@trigger.dev/database:generate:  ELIFECYCLE  Command failed.
command (/home/seven/Documents/codes/oss/trigger.dev/packages/database) pnpm run db:migrate:deploy exited (1)

 Tasks:    0 successful, 2 total
Cached:    0 cached, 2 total
  Time:    4.076s 
Failed:    @trigger.dev/database#db:migrate:deploy

 ERROR  run failed: command  exited (1)
 ELIFECYCLE  Command failed with exit code 1.
 ```

**Comment by matt-aitken** [comment]
@kaf-lamed-beyt it looks like Docker isn't running.

You probably just need to run `pnpm run docker`

It's step 6 in the setup in the Contributing guide: https://github.com/triggerdotdev/trigger.dev/blob/main/CONTRIBUTING.md

**Comment by kaf-lamed-beyt** [comment]
I actually ran docker the first time and it outputed a text similar to: "containers created" and exited.

But, I'll try doing it again, and share what I see.

**Comment by kaf-lamed-beyt** [comment]
This is what I'm seeing @matt-aitken

```shell
sudo pnpm run docker

> triggerdotdev@0.0.1 docker /home/seven/Documents/codes/oss/trigger.dev
> docker-compose -p triggerdotdev-docker -f docker/docker-compose.yml up -d

[+] Running 1/0
 ✔ Container database  Started                       
 ```

**Comment by matt-aitken** [comment]
@kaf-lamed-beyt it looks like you need to delete your docker volume that's clashing. In the Docker UI you can go to the volumes and delete them.

File: packages/cli/src/commands/dev.ts [hunk 1]
```diff
@@ -328,6 +330,77 @@ async function refreshEndpoint(apiClient: TriggerApi, endpointId: string, endpoi
...
+
+  if (response.ok) {
+    const data = await response.json();
+    // use array destructuring because of the `dist-tags` property
+    // @ts-ignore
+    return data["dist-tags"].latest;
+  } else {
+    console.log(`Failed to fetch the latest version of ${packageName}`);
+  }
+}
```
Comment (matt-aitken) [inline packages/cli/src/commands/dev.ts]: It'd be nice to add a bit more safety to this:

```ts
const packageRegisterSchema = z.object({
  "dist-tags": z.object({
    latest: z.string(),
  }),
});

async function getLatestPackageVersion(packageName: string) {
  const npmRegistryURL = `https://registry.npmjs.org/${packageName}`;
  try {
    const response = await fetch(npmRegistryURL);

    if (response.ok) {
      const data = await response.json();
      const parsedData = packageRegisterSchema.safeParse(data);

      if (parsedData.success) {
        return parsedData.data["dist-tags"].latest;
      }
    }
  } catch (e) {}

  console.log(`Failed to fetch the latest version of ${packageName}`);
}
```

File: packages/cli/src/commands/dev.ts [hunk 2]
```diff
@@ -328,6 +330,77 @@ async function refreshEndpoint(apiClient: TriggerApi, endpointId: string, endpoi
   }
 }
 
+async function getOutdatedTriggerDevPackages(path: string) {
+  const packageJSONPath = pathModule.join(path, "package.json");
+  const packageJSON = await readJSONFile(packageJSONPath);
+
+  const packageJSONSchema = z.object({
+    dependencies: z.record(z.string()),
+    devDependencies: z.record(z.string()),
```
Comment (matt-aitken) [inline packages/cli/src/commands/dev.ts]: `devDependencies` should be optional. The default with a new nextjs app is none

File: packages/cli/src/commands/dev.ts [hunk 3]
```diff
@@ -328,6 +330,77 @@ async function refreshEndpoint(apiClient: TriggerApi, endpointId: string, endpoi
...
+    }
+  }
+
+  if (outdatedPackages.length > 0) {
+    console.warn("⚠️ Outdated @trigger.dev packages found:");
+
+    for (const pkg of outdatedPackages) {
+      console.warn(`${pkg.name}: installed ${pkg.installedVersion}, latest ${pkg.latestVersion}`);
+    }
+    console.warn("Run 'npx @trigger.dev/cli update' to update the packages.");
```
Comment (matt-aitken) [inline packages/cli/src/commands/dev.ts]: For now, this should be commented out, as it'll be a different PR to do this.

File: packages/cli/src/commands/dev.ts [hunk 4]
```diff
@@ -328,6 +330,77 @@ async function refreshEndpoint(apiClient: TriggerApi, endpointId: string, endpoi
...
+  } else {
+    console.log("All @trigger.dev packages are up-to-date.");
```
Comment (matt-aitken) [inline packages/cli/src/commands/dev.ts]: If they're up to date I think it'd be better to log nothing

**Review by matt-aitken — CHANGES_REQUESTED** [review]

File: packages/cli/src/commands/dev.ts (hunk 4 above)
Comment (kaf-lamed-beyt) [reply packages/cli/src/commands/dev.ts]: Oh! Cool! I'll fix this now

File: packages/cli/src/commands/dev.ts (hunk 2 above)
Comment (kaf-lamed-beyt) [reply packages/cli/src/commands/dev.ts]: Cool. Noted!

File: packages/cli/src/commands/dev.ts (hunk 1 above)
Comment (kaf-lamed-beyt) [reply packages/cli/src/commands/dev.ts]: Ha! This looks better. I'll include it.

**Comment by kaf-lamed-beyt** [comment]
@matt-aitken I push some changes a moment ago. Kindly review.

I'm currently downloading the docker desktop app.
image

But, aside from that. I've also done 👇🏼 to remove unused volumes

```shell
sudo docker system prune
```
Currently, when I run 👇🏼 and try to delete it with the `rm` command like so: `sudo docker volume rm triggerdotdev-docker_database-data` I'd find out that it is still running. 

```shell
sudo docker volume ls
```

For now, I'm still looking for a workaround. When I find one, I'll let you know.

**Comment by kaf-lamed-beyt** [comment]
So awesome!! 🚀 

cc: @matt-aitken 

image

**Comment by kaf-lamed-beyt** [comment]
Hi @ericallam, @matt-aitken. 

In case you missed this.

**Comment by ericallam** [comment]
Can you double check this works? The screenshot you linked above looks like it's picking up an outdated package that isn't actually outdated:
CleanShot 2023-08-11 at 10 18 18

**Comment by kaf-lamed-beyt** [comment]
Okay, I'll check again now.

**Comment by kaf-lamed-beyt** [comment]
@ericallam, the reason why both versions kept on showing up was because I was comparing the version strings instead using the appropriate approach. 

So I used semver instead. 

image

**Comment by kaf-lamed-beyt** [comment]
@ericallam, I've included the changeset for this PR, kindly review.

**Comment by kaf-lamed-beyt** [comment]
@matt-aitken, whenever you get the chance, please help me review this.

File: packages/cli/src/commands/dev.ts [hunk 5]
```diff
@@ -359,6 +362,95 @@ async function refreshEndpoint(apiClient: TriggerApi, endpointId: string, endpoi
...
+  for (const packageName of triggerDevPackages) {
+    const latestVersion = await getLatestPackageVersion(packageName);
+    const installedVersion =
+      packageJSON.dependencies?.[packageName] || packageJSON.devDependencies?.[packageName];
+
+    // if latestVersion isn't undefined
+    // format the versions if they include special characters like "^", so semver can parse them
+    if (latestVersion) {
+      const formtattedInstalledVersion = semver.clean(installedVersion);
+      const formattedLatestVersion = semver.clean(latestVersion);
```
Comment (matt-aitken) [inline packages/cli/src/commands/dev.ts:428]: This code doesn't work with `^`.

"^2.0.8" is being turned into `null` which means that no packages are ever considered outdated.

Follow this guide on testing the CLI: https://github.com/triggerdotdev/trigger.dev/blob/main/CONTRIBUTING.md#testing-cli-changes

Test:
1. That it doesn't initially identify any outdated packages (i.e. does nothing).
2. Modify one of the versions to be older, then check if identifies it and logs it out

**Review by matt-aitken — CHANGES_REQUESTED** [review]

File: packages/cli/src/commands/dev.ts (hunk 5 above)
Comment (kaf-lamed-beyt) [reply packages/cli/src/commands/dev.ts:428]: @matt-aitken 

First off... if we're to modify the package version by editing the string to include `^`, with what I've observed it doesn't change the package version unless the developer does `npm install` again. 

And that means we may just be comparing the string values, and sometimes, before doing `npm install` the dev command will return that the packages (outdated and latest) are outdated as seen in this [comment](https://github.com/triggerdotdev/trigger.dev/pull/270#issuecomment-1674443817)

But, with semver, since that is what I used, all package with their key value pairs (i.e. the name and versions) are most times represented with the leading `^` and semver considers those strings as invalid package versions because I'm trying to compare them with their respective string values from the npm API response. 

So to avoid this issue, of invalid package versions, semver provides a `clean()` method to remove the `^` so that it can be considered as a valid one.

You can try modifying the code to exclude the check (with the `clean` method) and see the error I talked about.

Secondly, I'm sort of away from my computer for the next three weeks, I won't be able to access my PC till September 6 or thereafter. 

So, if that's okay, I'd still love to work on this issue by then, if my approach with semver doesn't quite fit the problem.

Comment (matt-aitken) [reply packages/cli/src/commands/dev.ts:428]: Hey @kaf-lamed-beyt,

Thanks for this, that makes sense. Now I understand why more frameworks don't do what we're trying to do here, it's not as simple as I thought it would be…

I think a better approach is to use this package: https://www.npmjs.com/package/npm-check-updates

It has a lot of contributors and high usage and allows programmatic usage (see the bottom of the readme) with wildcards.

If run from the command line:
```
npx npm-check-updates "/trigger.dev\/.+$/"
```

It correctly outputs
CleanShot 2023-08-18 at 11 42 33

When you're back if you can make this change we'll tip you generously because we've messed you around a bit here.

Comment (kaf-lamed-beyt) [reply packages/cli/src/commands/dev.ts:428]: Oh yeah! I've seen `npm-check-updates` in action – frequently in some of the projects I use.

I'll be happy to implement this when I'm back. Thank you!

**Comment by kaf-lamed-beyt** [comment]
Sure thing @ericallam! I'll do so. 

Thank you. ✊🏽

**Comment by kaf-lamed-beyt** [comment]
@ericallam, I noticed that @neo773 submitted a PR (#412) that was merged.

There won't be a need to work on this anymore, yeah? Per this [comment](https://github.com/triggerdotdev/trigger.dev/pull/270#issuecomment-1686514546)
